pubsub/gochannel/fanout.go

```diff
@@ -0,0 +1,99 @@
+package gochannel
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+)
+
+// FanOut is a component that receives messages from the subscriber and passes them
+// to all publishers. In effect, messages are "multiplied".
+//
+// A typical use case for using FanOut is having one external subscription and multiple workers
+// inside the process.
+//
+// You need to call AddSubscription method for all topics that you want to listen to.
+// This needs to be done *before* the router is started.
+//
+// FanOut exposes the standard Subscriber interface.
+type FanOut struct {
+	internalPubSub *GoChannel
+
+	router     *message.Router
+	subscriber message.Subscriber
+
+	logger watermill.LoggerAdapter
+
+	subscribedTopics map[string]struct{}
+	subscribedLock   sync.Mutex
+}
+
+// NewFanOut creates new FanOut.
+// The passed router should not be running yet.
+func NewFanOut(
+	router *message.Router,
+	subscriber message.Subscriber,
+	logger watermill.LoggerAdapter,
+) (*FanOut, error) {
+	if router == nil {
```

Comment:
basically, is there any case when we need to pass router?

Reply:
Moved to internal router in a96ad451dc7bc6c5258ef8c684dbc6d7270af7a4